File: advent_of_code/2018/day2/inventory-management-system.js
// Part 1

const { readFileSync } = require('fs')
const path = require('path')

function calculateChecksum(file) {
  let ids = readFileSync(file).toString().split('\n')
  let total = [0, 0]

  for (let line of ids) {
    let letters = {}
    for (let c of line) {
      if (letters[c]) letters[c] += 1
      else letters[c] = 1
    }
    if (Object.values(letters).includes(2)) total[0] += 1
    if (Object.values(letters).includes(3)) total[1] += 1
  }

  return total[0] * total[1]
}

let checksum = calculateChecksum(path.join(__dirname, 'day2.txt'))
console.log(checksum)

// Part 2

function commonLettersCorrectIDs(file) {

}

let commonLetters = commonLettersCorrectIDs(file)
console.log(commonLetters)

File: advent_of_code/2018/day2/inventory-management-system.py
import os
from collections import Counter

# Part 1

def calculate_checksum(file):
    c2 = 0
    c3 = 0

    with open(file, 'r') as f:
        for line in f:
            line = line.strip()
            letter_counts = Counter(line).values()
            if 2 in letter_counts:
                c2 += 1
            if 3 in letter_counts:
                c3 += 1

    return c2 * c3

dir_path = os.path.dirname(os.path.abspath(__file__))
input_file = os.path.join(dir_path, 'day2.txt')

checksum = calculate_checksum(input_file)
print(checksum)

# Part 2

def common_letters_correct_ids(file):
    with open(file, 'r') as f:
        lines = f.read().strip().splitlines()

    for i in lines:
        for j in lines:
            diff_count = 0
            diff_index = None
            for k, (c1, c2) in enumerate(zip(i, j)):
                if c1 != c2:
                    diff_count += 1
                    diff_index = k
                if diff_count > 1:
                    break
            if diff_count == 1:
                return i[:diff_index] + i[diff_index+1:]

    return None

common_letters = common_letters_correct_ids(input_file)
print(common_letters)
